### ISSUE TITLE
Sigmoid causing Python to hang

### ISSUE THREAD

**Issue by twesterhout** [pytorch/pytorch#17225]
## 🐛 Bug

`torch.sigmoid` causes python to hang when used from `torch.jit.ScriptModule` together with `torch.multiprocessing`.

## To Reproduce

The following simple script should reproduce the issue:

```python
import torch
from torch.multiprocessing import Process

@torch.jit.script
def sigmoid(x):
    return 1.0 / (1.0 + torch.exp(-x))


class Net(torch.jit.ScriptModule):
    def __init__(self, n):
        super().__init__()

    @torch.jit.script_method
    def forward(self, x):
        return torch.sigmoid(x)
        # return sigmoid(x)
        # return torch.tanh(x)

def run(rank,):
    xs = torch.randn((5, 3), dtype=torch.float32)
    f = Net(3)
    print("{}: Starting...".format(rank))
    print(f.forward(xs))
    print("{}: Done!".format(rank))

if __name__ == "__main__":
    # These lines are important
    xs = torch.randn((5, 3), dtype=torch.float32)
    net = Net(3)
    print(net(xs))

    processes = []
    for rank in range(2):
        p = Process(target=run, args=(rank,))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

The result is:
```
tensor([...])
0: Starting...
1: Starting...
```
and the process hangs.

Now if we use `torch.tanh` in place of `torch.sigmoid`, all works well:
```
tensor([...])
0: Starting...
1: Starting...
tensor([...])
0: Done!
tensor([...])
1: Done!
```
the order of messages may differ, of course, but the important part is that the script finishes successfully. Using a dummy implementation of `sigmoid` helps avoid the issue as well.

## Expected behavior

For the script to finish.

## Environment

`collect_env.py` output:
```
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Debian GNU/Linux buster/sid
GCC version: (Debian 8.2.0-16) 8.2.0
CMake version: version 3.13.4

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.15.4
[pip3] torch==1.0.1.post2
[conda] Could not collect
```

**Comment by driazati** [comment]
This doesn't look like it's a jit issue, it also happens on a plain `torch.nn.Module`

**Comment by colesbury** [comment]
Threads and fork do not mix well. Use "spawn" or "fork" multiprocessing modes or avoid initializing the OpenMP thread pool via `print(net(xs))` before you launch your subprocesses.

https://docs.python.org/3/library/multiprocessing.html#multiprocessing.set_start_method

The OpenMP threads pool is initialized in the first `print(net(xs))`. Threads are not inherited by a fork'd process, but the state associated with them is. That's the underlying cause of the deadlock.

**Comment by twesterhout** [comment]
Thanks @colesbury! You're right. Is there a way to detect at runtime that OpenMP thread pool has been initialised and throw on attempts to fork? A big warning sign in the docs of `torch.multiprocessing` would also help.

**Comment by colesbury** [comment]
@twesterhout , I don't know of a way to detect if the OpenMP thread pool was initialized. I don't think that's part of the OpenMP API. 

@lerks mentions a work-around which is to `torch.set_num_threads(1)` before the fork.

Let's continue discussion about possible improvements in #17199
